### DISCUSSION TITLE
Does Github Pages support HSTS TLDs for custom domains?

### DISCUSSION THREAD

**Discussion by nuxy** [community/community#113716] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I'm trying to use `.dev` TLD for my Github [project pages](https://textmarked.nuxy.dev), and while Github [successfully verifies the domain name](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/verifying-your-custom-domain-for-github-pages), the corresponding Github page returns a web browser initiated HSTS error message. I have not been able to find [anything documented](https://github.com/search?q=repo%3Agithub%2Fdocs%20hsts&type=code) regarding this HSTS-required domain names.

A definitive answer from Microsoft would be helpful in this regard, or update to the [Pages documentation](https://github.com/github/docs/blob/main/content/pages/configuring-a-custom-domain-for-your-github-pages-site/about-custom-domains-and-github-pages.md) that clarifies HSTS support.

**Comment by yoannchaudet** [comment]
I am not aware of us not supporting HSTS TLDs (like `.dev`). If you share your domain name, we can look up what may be preventing the cert obtention.

**Comment by Tgentil** [comment, marked as answer]
Yes, GitHub Pages supports custom domains, including those requiring HSTS like `.dev` domains. When you configure a custom domain on GitHub Pages, you can enforce HTTPS, which is essential for HSTS compliance. It's important to correctly set up DNS records with your DNS provider for this purpose. 

For detailed guidance on setting up your custom domain and enforcing HTTPS on GitHub Pages, you might find these GitHub Docs helpful:
- [Managing a custom domain for your GitHub Pages site](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/about-custom-domains-and-github-pages)
- [Securing your GitHub Pages site with HTTPS](https://docs.github.com/en/pages/getting-started-with-github-pages/securing-your-github-pages-site-with-https)

I also found this blog very helpful:

- [Connecting a .DEV Domain to Github Pages with HTTPS](https://medium.com/@jaredpotter1/connecting-a-dev-domain-to-github-pages-with-https-f0c59fb62579)

**Comment by nuxy** [comment]
This seems to be related to an issue on my part when Github Pages was set-up in conjuction with [verifying custom domains](https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/verifying-your-custom-domain-for-github-pages]verifying-your-custom-domain), and quite possibly DNS record caching. After falling back to using only the CNAME the page can now be accessed successfully.

**Comment by RyanCavanaugh** [comment]
@Tgentil 's answer above is incomplete. GitHub Pages supports HTTPS, and will send HSTS *for domains that require it*, e.g. `.dev`. However, there is currently not a way to enable HSTS for domains which *don't* require HSTS, e.g. `.com` or `.org`.

**Comment by UnoYakshi** [comment]
So there is no built-in option to turn on HSTS for GHP?

**Reply by RyanCavanaugh** [reply]
You can email Github Support and they can enable it for a particular custom domain

**Reply by UnoYakshi** [reply]
I will try it, thank you.

**Reply by UnoYakshi** [reply]
It works, indeed. However, support can't help setting other security headers. A shame for 2025.

**Reply by ElioDiNino** [reply]
Also confirming that contacting support to enable HSTS for a particular domain works!

**Reply by cicirello** [reply]
@ElioDiNino any idea if you need to request the HSTS header for specific subdomains? if it is done right, it should be served for any subdomains of a domain. it looks like they have it for both your status and internal subdomains. Did you specifically request that? The reason that I ask is looking at the headers for another in this thread (@UnoYakshi ), the header is served with the root domain, but not with the www subdomain. Although in that case, it just forwards to the root domain, I think the header for that forward should probably include the HSTS header as well. Anyway, I have several project sites on subdomains, and just looking to know if I need to explicitly ask support for each of the subdomains or just for the root.

**Reply by ElioDiNino** [reply]
@cicirello Yes, I specifically requested it for each site. My message was as follows:

> Per https://github.com/orgs/community/discussions/113716#discussioncomment-12344699, I am requesting HSTS be enabled on the following GitHub Pages sites that I own:
> * status.eliodinino.com (repo: https://github.com/ElioDiNino/status.eliodinino.com)
> * internal.eliodinino.com (repo: https://github.com/ElioDiNino/internal.eliodinino.com)
> * mta-sts.eliodinino.com (repo: https://github.com/ElioDiNino/mta-sts.eliodinino.com)
> * health.eliodinino.com (repo: https://github.com/ElioDiNino/health.eliodinino.com)

Their reply was (to which I replied yes to both):

> Thanks for reaching out to GitHub Support to request that we enable HSTS for your custom domain . I would be happy to help you with this request.
>
> Do you wish to enable the preload directive?
> 
> Do you wish to include subdomains in the HSTS policy?

**Reply by cicirello** [reply]
Thanks @ElioDiNino
